### ISSUE TITLE
Should we remove a bunch of compile time `-D` flags?

### ISSUE THREAD

**Issue by krader1961** [att/ast#210]
I've started working on restyling the ksh source. It's amazing how many `#if SOMETHING` pragmas there are and how much harder it makes it to read the code. Since many of them are always defined (see `ksh93_c_args` in *src/cmd/ksh93/meson.build*) I propose we simply drop those symbols and removing the `#if` pragmas that reference them. It's possible a couple of them should remain so people could, theoretically, build a ksh without the feature. But it looks to me like most of them should always be enabled.

**Comment by siteshwar** [comment]
Yep, I noticed it too and most of these ```#if```'s should be removed.

**Comment by krader1961** [comment]
To clarify my opening comment.... I don't see anyone ever compiling it without `-DSHOPT_MULTIBYTE` for example. Or `-DSHOPT_BASH` and `-DSHOPT_KIA`. The only one that vaguely looks like it might make sense to remain optional is `-DSHOPT_COSHELL`.

I'll send an email to the ast-developers mailing list as well but if anyone lurking can explain why it should be possible to build ksh without these features that would be helpful. Alternatively, if you strongly believe these conditional features should always be included say so.

**Comment by DavidMorano** [comment]
> To clarify my opening comment.... I don't see anyone ever compiling it without -DSHOPT_MULTIBYTE for
>  example. Or -DSHOPT_BASH and -DSHOPT_KIA. The only one that vaguely looks like it might make 
> sense to remain optional is -DSHOPT_COSHEL

I do not know about all of these flags in  detail (off-hand), but I have had to turn OFF more than
one of these in order to get the bugs out within the optional code (on some platforms).
Among the very first of these options often needing to be turned OFF is -DSHOPT_MULTIBYTE !
Just my experience.  In fact, I wish that *more* of the whole of KSH was more optioned so that
more of the bugs can be optioned out.  For example, the "sophisticated" command line completion
and command line multi-line display are very buggy (causing core dumps) on some platforms.
I wish they could be optioned out!

In a few cases, I added compile-time options of my own in order to get rid of some code that was
causing bad bugs.  Just to be clear: "bad" bugs are often core dumps!

**Comment by siteshwar** [comment]
It would be best to have such bugs reported and fixed. Working around bugs by trimming down features is not the option.

**Comment by krader1961** [comment]
Thank you, @DavidMorano, for taking the time to comment. However, I agree with @siteshwar. If a feature is so broken that it causes people like you to want to build ksh without the feature then the feature has no business being included in a stable release. We should not be hiding broken features behind build time flags. With the possible exception of the bleeding edge development branch and even then the flag should not exist for more than a few months. Once a feature is ready to be included in a stable release it should no longer be guarded by a compile time `#define`.

I definitely share your concerns, @DavidMorano, about the quality of a lot of the recently added code. That is, code added since version 93u+, released 2012-08-01, which seems to be the basis of the `ksh` binary included in every distro I use. But the answer is not to complicate the code with `#if SOME_SYMBOL` tests. The answer is to either fix those bugs or remove the feature.

**Comment by krader1961** [comment]
@DavidMorano,

Your comment suggests that you have actually looked at, and attempted to modify, the ksh source code. Is that correct? I ask because I am extremely unhappy about the current number of blocks of code conditioned on `#if SOME_SYMBOL... #endif` tests. It makes the code extremely hard to read, format, run through tools like `oclint`, and test. I've been programming in C, including OS kernel code (not just Linux), for three decades. I've never seen a project that makes such liberal use of conditional blocks of code outside of *.h* header files. So I'm perplexed by your comment where you seem to suggest there should be more instances of that anti-pattern.

**Comment by DavidMorano** [comment]
> Your comment suggests that you have actually looked at, and attempted to modify, 
> the ksh source code. Is that correct?
> ...

I hear you and sympathize with where you are coming from.

Yes, that is correct. I confess. I apologize for causing some extra concern
here. I have at times gone through the KSH source (more than I ever wanted to)
to compile-time disable code either by an existing feature switch or by inserted
compile-time switches of my own (sometimes possibly compiling in a little
alternate code) in order to get around some problem. I confess that I usually do
not have the time to properly figure out exactly where the bugs lie. Instead my
(admittedly) selfish goal was to get a working compiled KSH as fast as possible.
I wish with all of my heart that someone paid me to clean up KSH (get all of the
bugs out), but sadly that is not the case. To this day, I live every day with a
KSH that crashes (core dumps) if I do something that irritates it in just the
wrong way. But I live with this because I understand that both my resources and
those of the open-source community in general are quite limited. In fact, this
applies to much of all the software I use everyday, open-source or commercial.
There are bugs in almost every piece of software I use now-a-days. Somehow --
and I certainly wish this was *not* the case -- I (being especially cursed it
seems) find bugs in much of the the software I use regularly. I think this is
just the world we live in now.

KSH is complicated enough -- at least for me -- such that attempting to properly
track down some of its bugs would likely be quite time consuming. For example,
there is a very subtle I/O redirection related or associated bug that has been
in there for about fifteen years now, but it is subtle enough to live with it
rather than dedicating my life towards finding and correcting it. There are some
subtle esoteric SFIO bugs that I would like see cleaned up (again in there for
over fifteen or more years) , but good luck with that. For me, the SFIO code is
not the easiest to always follow, and as I stated already, I do not have the
time to become the next SFIO guru for planet Earth. So I live with this stuff as
it is. Again, sorry for being so selfish.

My suggestion was not to necessarily keep all of the compile-time switches in
the code base, but rather to retain an ability to disable features of the code
(like multibyte, or command completion, or multi-line command display, or ...
the whole VMALLOC facility for that matter, and more) at run-time, when they
turn up in a so-called "stable" release, but are either malfunctioning or
outright broken to the point of core dumping on some platforms! There could be
run-time switches in the code that are set either by option key-words in one or
more environment variables or by additional options to the KSH 'set' command
(for example). But something to switch out some of the supposedly fancy features
that do not automatically work properly on all platforms would still be
welcomed.

For myself, I have used KSH just about every day for the last 34 years (nearly
from its very inception). I have written over 110 KSH builtin commands (many of
them being huge codes, using numerous dynamically loaded shared-object
libraries, and application loaded shared objects) not to mention functions and
scripts. I have KSH built-ins that are multithreaded. I have background daemon
threads or server threads running along side KSH in its own background. So few
people want to see KSH stay around more than I do, but the most important thing
for widespread adoption is for an end user-developer-local-maintainer or
whatever, to get the code base and get it complied to a point where it is stable
enough to use as quickly as possible. We should not expect everyone who
downloads the KSH source base to debug every bug they find before they are
allowed to reasonably use it as a tool. Having something like some feature
switches (implemented in whatever way) might help with the local adoption
process by possibly disabling code functions which a) are not strictly needed,
and b) might be malfunctioning, or even c) might be a feature which the local
organization finds either unwelcomed behavior or outright abhorrent.

Is the KSH source base messy? YES. But just as important or even more so, there
are numerous bugs, subtle and not so subtle, in the code base. But simply
removing the existing compile-time switches is not going to remove those bugs.
Should the bugs be removed? Of course. How that should be best done I will leave
to others. I wish I had the time to rewrite much of the KSH base from
essentially the ground up (not the place here to elaborate on alternate code
architectures). But I do not see that happening any time soon.

I was only pointing out that having features that can be disabled has proven
useful at times in getting rid of code that has malfunctioned on one or more
platforms.

**Comment by krader1961** [comment]
FWIW, @DavidMorano, my intention is to remove dependencies on things like the AST vmalloc and SFIO subsystems once we've made the build process and unit tests more robust and less flakey. I have no interest in maintaining that code and, as you've noted, they are likely to be the source of bugs. Since the POSIX/Xopen standards provide the functionality needed by ksh and every platform we expect to run on provide support for those standards it doesn't make sense to maintain our own implementations of those APIs.

**Comment by DavidMorano** [comment]
@krader1961 
> my intention is to remove dependencies on things like the AST vmalloc and SFIO 
> ...

I would welcome VMALOC being removed, 
and all of the other AST specific things that were originally designed to stub out
for missing POSIX/Xopen  interfaces. I found that I had to remove VMALLOC
because even though VMALLOC had a compile-time option to make it 
thread-safe, for some reason it was not thread-enough-safe even in so-called
"thread-safe" compile mode (had races and core dumps).  Also, for me, I would 
welcome the removal of SFIO!  Yes, I said that correctly.  I actually find SFIO quite 
deficient for my needs and I do not want to take the plunge to go into the code
to learn enough about it to make it more suitable for my needs (mostly related to
asynchronous operations).

Currently, KSH mandates the use of SFIO for its I/O interface with any
coded built-ins.  This has been a huge problem for me in the past, and
continuing, because of either bugs or deficiencies within SFIO.
So I would welcome the removal of SFIO just so the standard I/O within 
built-ins can be released from their present bondage.

Built-ins that people may have which wish to continue using SFIO
could continue to do so, but the way those streams would get initialized would
work differently.  Currently SFIO streams for STDIN, STDOUT, and STDERR
are inherited from KSH itself rather than  being initialized from the
corresponding FDs (STDIN, STDOUT, and STDERR).  Korn did this
so that purely memory based streams could be passed into and out
of built-ins without actually using any FDs. But with the removal of
SFIO from KSH, built-ins which continued to use SFIO, would have to
have their standard streams (IN, OUT, and ERR) initialized from the
corresponding FDs (IN, OUT, and ERR) just like STDIO (and others)
do when running in the context of a newly spawned process.

So I do not see any real problem with the removal of SFIO, but I do
see up-sides to its removal (bugs gone and being released from its constrained
interface and semantics).

In short, I would welcome all of these sorts of changes.
I only now wish that this stuff was removed at least 20 years ago,
since it would have saved me innumerable head and heart aches working 
around them as I had to do (too numerous to elaborate on in this space).

Thank you very much for your plan.  I assume that you have the time do all
of this. so thanks for that also!

**Comment by krader1961** [comment]
Note that disabling features can cause compilation failures. For example, compiling with `-DSHOPT_MULTIBYTE=0` results in this error:

```
../src/cmd/ksh93/sh/string.c:348:23: error: ‘lc_unicodeliterals’ undeclared (first use in
  if(isaletter(c) && (!lc_unicodeliterals || c<=0x7f))
                       ^
```

**Comment by krader1961** [comment]
While restyling the code I am amazed at how many `#if SOME_SYMBOL` directives (or equivalent) I am seeing. For example, `SHOPT_SYSRC`, `ATTRS` and `STR_MAXIMAL`. Just to pick three at random. I recognize that conditionally compiled code is necessary to support really old, as well as new, systems. But the manner in which this has been done is suboptimal and needlessly confusing. And it appears the majority of these tests are for things that should not be conditionally compiled and are unrelated to features available on a specific system. Which is the nicest way I can say what I really think. Ask me out for a beer for a more blunt appraisal of the state of the code.

**Comment by krader1961** [comment]
FWIW, The command I ran against the source to find all the compile time symbols looks like this:

```
grep -P -h '^#\s*if' ***.c ***.h | sort | uniq -c
```

The output of that command is 2062 lines. Which is astonishing for the source of a project like this. The output for `zsh`, for example, is only 716 lines as I write this. For `bash` it is 1752 lines. Which is almost as bad and likely reflects the fact it too is trying to accommodate long dead environments.

**Comment by krader1961** [comment]
Also for your amusement is this compiler directive in *src/lib/libcmd/grep.c*:

```
#if _you_like_the_noise
                error(1, "newline appended");
#endif
```

WTF!

**Comment by krader1961** [comment]
I've taken care of all the easy `SHOPT_`&ast; cleanups. There are some still to be resolved (e.g., `SHOPT_ENV` and `SHOPT_BASH`). Those have separate issues. Closing since this issue has been substantially resolved.
